chain/index/api.go

```diff
@@ -126,17 +127,124 @@ func (si *SqliteIndexer) ChainValidateIndex(ctx context.Context, epoch abi.Chain
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get tipset message and event counts at height %d: %w", expectedTs.Height(), err)
+	}
+
+	if err = si.verifyIndexedData(ctx, expectedTs, indexedData); err != nil {
+		return nil, xerrors.Errorf("failed to verify indexed data at height %d: %w", expectedTs.Height(), err)
+	}
+
 	return &types.IndexValidation{
-		TipsetKey: expectedTs.Key().String(),
-		Height:    uint64(expectedTs.Height()),
-		// TODO Other fields
+		TipsetKey:      expectedTs.Key().String(),
+		Height:         uint64(expectedTs.Height()),
+		TotalMessages:  uint64(indexedData.nonRevertedMessageCount),
+		TotalEvents:    uint64(indexedData.nonRevertedEventCount),
+		EventsReverted: indexedData.hasRevertedEvents,
 	}, nil
 }
 
+type indexedTipSetData struct {
+	nonRevertedMessageCount int
+	nonRevertedEventCount   int
+	hasRevertedEvents       bool
+}
+
+// getIndexedTipSetData fetches the indexed tipset data for a tipset
+func (si *SqliteIndexer) getIndexedTipSetData(ctx context.Context, tsKey types.TipSetKey) (*indexedTipSetData, error) {
+	cid, err := tsKey.Cid()
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get tipset key cid: %w", err)
+	}
+	tsKeyBytes := cid.Bytes()
+
+	var data indexedTipSetData
+	err = withTx(ctx, si.db, func(tx *sql.Tx) error {
+		if err = tx.Stmt(si.stmts.getNonRevertedTipsetMessageCountStmt).QueryRowContext(ctx, tsKeyBytes).Scan(&data.nonRevertedMessageCount); err != nil {
+			return xerrors.Errorf("failed to query non reverted message count: %w", err)
+		}
+
+		if err = tx.Stmt(si.stmts.getNonRevertedTipsetEventCountStmt).QueryRowContext(ctx, tsKeyBytes).Scan(&data.nonRevertedEventCount); err != nil {
+			return xerrors.Errorf("failed to query non reverted event count: %w", err)
+		}
+
+		// we are only fetching non-reverted events, so if there are no non-reverted events,
+		// then we need to check if the events were reverted.
+		if data.nonRevertedEventCount == 0 {
+			if err = tx.Stmt(si.stmts.hasRevertedEventsStmt).QueryRowContext(ctx, tsKeyBytes).Scan(&data.hasRevertedEvents); err != nil {
+				return xerrors.Errorf("failed to check for reverted events: %w", err)
+			}
+		}
+
+		return nil
+	})
+
+	return &data, err
+}
+
+// verifyIndexedData verifies that the indexed data for a tipset is correct
+// by comparing the number of messages and events in the chainstore to the number of messages and events indexed.
+// NOTE: Events are loaded from the executed messages of the tipset at the next epoch (ts.Height() + 1).
+func (si *SqliteIndexer) verifyIndexedData(ctx context.Context, ts *types.TipSet, indexedData *indexedTipSetData) (err error) {
+	tsKeyCid, err := ts.Key().Cid()
+	if err != nil {
+		return xerrors.Errorf("failed to get tipset key cid: %w", err)
+	}
+
+	msgs, err := si.cs.MessagesForTipset(ctx, ts)
+	if err != nil {
+		return xerrors.Errorf("failed to get messages for tipset: %w", err)
+	}
+
+	msgCount := len(msgs)
+	if msgCount != indexedData.nonRevertedMessageCount {
```

Comment:
You don't need to call `MessagesForTipset` here and compare number of messages since you are calling `loadExecutedMessages` below, right ? You can just compare then.

Reply:
Look at what the `loadExecutedMessages ` in it's implementation. Reading the implementation code will give you a better  understanding of how it all works together.

Reply:
Just checked, you're correct it's the same thing, I will update.

Reply:
Done.